src/Microsoft.DotNet.Interactive/KernelSupportsNugetExtensions.cs

```diff
@@ -178,36 +166,58 @@ public static string GetDisplayValueId(PackageReference package)
             public static IEqualityComparer<PackageReference> Instance { get; } = new PackageReferenceComparer();
         }
 
+        private static void CreateOrUpdateDisplayValue(KernelInvocationContext context, string name, object content, string mimeType)
+        {
+            object displayed = null;
+            if (!context.Command.Properties.TryGetValue(name, out displayed))
+            {
+                displayed = context.Display(content, mimeType);
+                context.Command.Properties.Add(name, displayed);
+            }
+            else
+            {
+                (displayed as DisplayedValue).Update(content);
+            }
+        }
+
         internal static KernelCommandInvocation DoNugetRestore()
         {
             return async (command, invocationContext) =>
             {
                 KernelCommandInvocation restore = async (_, context) =>
                 {
+                    object InstallMessage(string message, IEnumerable<string> items, string progress = null)
+                    {
+                        if (items.Count() > 0)
+                        {
+                            return div(
+                                strong(message),
+                                ul(items.Select(s => li(span(s + progress)))));
+                        }
+                        else
+                            return div();
+                    }
+
                     if (!(context.HandlingKernel is ISupportNuget kernel))
                     {
                         return;
                     }
 
-                    var messages = new Dictionary<PackageReference, string>(new PackageReferenceComparer());
-                    var displayedValues = new Dictionary<string, DisplayedValue>();
-
                     var newlyRequestedPackages =
                             kernel.RequestedPackageReferences
                                           .Except(kernel.ResolvedPackageReferences, PackageReferenceComparer.Instance)
                                           .ToArray();
 
                     var requestedPackageIds = new Dictionary<string, PackageReference>();
+                    var progress = "    ";
 
-                    foreach (var package in newlyRequestedPackages)
-                    {
-                        var id = PackageReferenceComparer.GetDisplayValueId(package);
-                        var message = InstallingPackageMessage(package) + "...";
-                        var displayedValue = context.Display(message);
-                        displayedValues[id] = displayedValue;
-                        messages.Add(package, message);
-                        requestedPackageIds.Add(id, package);
-                    }
+                    object install = InstallMessage("Restore sources", (kernel.RestoreSources.OrderBy(s => s)));
```

Comment:
A `ToArray` to prevent multiple enumerations might be good here.

Reply:
done.